independent-projects/qute/core/src/main/java/io/quarkus/qute/JsonEscaper.java

```diff
@@ -1,10 +1,114 @@
 package io.quarkus.qute;
 
+import java.util.Arrays;
 import java.util.Optional;
 
 import io.quarkus.qute.TemplateNode.Origin;
 
-public class JsonEscaper extends CharReplacementResultMapper {
+public final class JsonEscaper implements ResultMapper {
+
+    private static final int NO_NEED_REPLACEMENT_DATA = 0x0100_0000;
+    // this int is packing:
+    // length of replacement (1 byte), padding byte,
+    // additional char (1 byte), replacement char (1 byte) = 4 bytes
+    private static final int[] REPLACEMENTS_DATA = new int[256];
+    private static final String[] CTRL_REPLACEMENTS = new String[32];
+
+    static {
+        // All Unicode characters may be placed within the quotation marks,
+        // except for the characters that MUST be escaped: quotation mark,
+        // reverse solidus, and the control characters (U+0000 through U+001F).
+        // See also https://datatracker.ietf.org/doc/html/rfc8259#autoid-10
+
+        // by default ctrl ASCII chars replace 6 chars
+        Arrays.fill(REPLACEMENTS_DATA, 0, 32, 0x0600_0000);
+        // default Latin chars just replace themselves
+        for (int i = 32; i < 256; i++) {
+            REPLACEMENTS_DATA[i] = i | (1 << 24);
+        }
+        // special ASCII chars - which include some control chars: replace 2 chars
+        REPLACEMENTS_DATA['"'] = (0x005c | (('"' & 0xFF) << 8)) | (2 << 24);
+        REPLACEMENTS_DATA['\\'] = (0x005c | (('\\') << 8)) | (2 << 24);
```

Comment:
To improve code clarity and maintainability, consider defining constants to give context to the values 24, 2 and 8. These values might represent something like a heap size or a bit size. Providing descriptive names for these constants will help others understand the code more easily and avoid magic number violations.
```suggestion
        // special ASCII chars - which include some control chars: replace 2 chars
        REPLACEMENTS_DATA['"'] = (0x005c | (('"' & 0xFF) << BIT_SIZE)) | (TWO_CHARS << THREE_BIT_SIZE);
        REPLACEMENTS_DATA['\\'] = (0x005c | (('\\') << BIT_SIZE)) | (TWO_CHARS << THREE_BIT_SIZE);
```